_automation/main.go

```diff
@@ -46,15 +49,20 @@ func (g *Grammar) ContentURL() string {
 }
 
 func (g *Grammar) FetchNewVersion() *GrammarVersion {
-	if strings.HasPrefix(g.Reference, "v") {
-		tag, rev := fetchLastTag(g.URL)
+
+	//attempt to retrieve tag from latest RELEASE in this grammar's repo
```

Comment:
I think it is going to break something or at least would require changing the behavior again in not so distant future.

The current logic is:
- if `grammars.json` uses tag version try to find the latest tag
- otherwise use the latest commit

The reason for this logic is that some grammars don't cut releases in a timely manner. This causes project not being able to parse code with new language features. A good example would be Go grammar:
- v0.19.1 was released on Mar 4, 2021
- v0.20.0 on Jul 26, 2023

During that time a number of new versions of Go were released and while the grammar was getting support for the new features (e.g. Generics) we couldn't pull it here by relying on release tag only. 

I would prefer to keep this flexibility if possible.
Afaiu the problem you have faced is that Swift grammar repository doesn't prefix its releases with `v`.
I would propose to modify the `if` statement to use regexp like `v?\d+\.\d+\.\d+`. Sadly `git ls-remote` doesn't support regexp, only glob, so we would also need to update  `fetchLastTag` to iterate over all tags looking for any that matches `v?\d+\.\d+\.\d+`.

Reply:
> The reason for this logic is that some grammars don't cut releases in a timely manner. This causes project not being able to parse code with new language features. 

That makes a lot of sense. I ran into the opposite issue when updating [Kotlin](https://github.com/fwcd/tree-sitter-kotlin). If I pull the latest commit, it's actually broken. The last working version is the one tagged in the latest Release.

> I would prefer to keep this flexibility if possible.

I agree. I think the functionality should remain as-is and pull from the most recent commit. I'd like to *add* a property in `grammars.json` that would instead instruct the updater to use the commit from the latest Release for that specific grammar, like `"preferLatestRelease": true` For example, if we did it with Kotlin it would look like:

```yaml
  {
    "language": "kotlin",
    "url": "https://github.com/fwcd/tree-sitter-kotlin",
    "files": [
      "parser.c",
      "scanner.c"
    ],
    "reference": "main",
    "revision": "cce9073807c6c1da81ce658d683fdac2878e5b57",
    "preferLatestRelease": true,
  },```

What do you think of this as a solution? It would leave the current behavior as-is, and allow grammar's to opt-in to using the most recent Release commit instead of simply the most recent commit.

Swift also has another issue in that it doesn't provide the generated parser.c inside the repo like every other grammar does. Instead it creates a `generated-src.zip` file attached to a Release. I don't see a way to get around having to write a custom function for just Swift (as I've done), but I'm open to other ideas.

Reply:
> The reason for this logic is that some grammars don't cut releases in a timely manner. This causes project not being able to parse code with new language features. 

That makes a lot of sense. I ran into the opposite issue when updating [Kotlin](https://github.com/fwcd/tree-sitter-kotlin). If I pull the latest commit, it's actually broken. The last working version is the one tagged in the latest Release.

> I would prefer to keep this flexibility if possible.

I agree. I think the functionality should remain as-is and pull from the most recent commit. I'd like to *add* a property in `grammars.json` that would instead instruct the updater to use the commit from the latest Release for that specific grammar, like `"preferLatestRelease": true` For example, if we did it with Kotlin it would look like:

```yaml
  {
    "language": "kotlin",
    "url": "https://github.com/fwcd/tree-sitter-kotlin",
    "files": [
      "parser.c",
      "scanner.c"
    ],
    "reference": "main",
    "revision": "cce9073807c6c1da81ce658d683fdac2878e5b57",
    "preferLatestRelease": true,
  },
```

What do you think of this as a solution? It would leave the current behavior as-is, and allow grammar's to opt-in to using the most recent Release commit instead of simply the most recent commit.

Swift also has another issue in that it doesn't provide the generated parser.c inside the repo like every other grammar does. Instead it creates a `generated-src.zip` file attached to a Release. I don't see a way to get around having to write a custom function for just Swift (as I've done), but I'm open to other ideas.

Reply:
If you want to use a separate field in `grammars.json` instead of relying on format of the reference as I suggested, it is also fine. Though I would name it something like `updateBasedOn` with values: `tag`, `commit`.

So that existing grammars will be updated in the following way:
```js
[
  {
    "language": "bash",
    "url": "https://github.com/tree-sitter/tree-sitter-bash",
    "files": [
      "parser.c",
      "scanner.cc"
    ],
    "reference": "v0.19.0",
    "revision": "b6667bed364733c8a8f8e5973749f86cfa04ba2a",
    "updateBasedOn": "tag"
  },
  {
    "language": "groovy",
    "url": "https://github.com/murtaza64/tree-sitter-groovy",
    "files": [
      "parser.c",
      "scanner.c"
    ],
    "reference": "main",
    "revision": "235009aad0f580211fc12014bb0846c3910130c1",
    "updateBasedOn": "commit"
  },
...
]
```

For the zip issue, yes, we need to use a special `download` function that you introduced in this PR.

Reply:
I like this idea of `updateBasedOn` and have added it to my updated PR